src/main/java/com/epam/esm/service/GiftCertificateService.java

```diff
@@ -25,23 +26,28 @@
 public class GiftCertificateService {
 
     private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+
     private final GiftCertificateDao giftCertificateDao;
     private final GiftCertificateTagDao giftCertificateTagDao;
     private final TagDao tagDao;
+    private final OrderDao orderDao;
     private final GiftCertificateValidator validator;
     private final GiftCertificateDtoMapper dtoMapper;
+    private final Paginator paginator;
 
     @Autowired
-    public GiftCertificateService(GiftCertificateDao giftCertificateDao, GiftCertificateTagDao giftCertificateTagDao, TagDao tagDao, GiftCertificateValidator validator, GiftCertificateDtoMapper dtoMapper) {
+    public GiftCertificateService(GiftCertificateDao giftCertificateDao, GiftCertificateTagDao giftCertificateTagDao, TagDao tagDao, OrderDao orderDao, GiftCertificateValidator validator, GiftCertificateDtoMapper dtoMapper, Paginator paginator) {
         this.giftCertificateDao = giftCertificateDao;
         this.giftCertificateTagDao = giftCertificateTagDao;
         this.tagDao = tagDao;
+        this.orderDao = orderDao;
         this.validator = validator;
         this.dtoMapper = dtoMapper;
+        this.paginator = paginator;
     }
 
-    public List<GiftCertificateDto> getGiftCertificates(String tagName, String giftCertificateName, String description, String sortByName, String sortByDate) {
-        return giftCertificateDao.getGiftCertificates(tagName, giftCertificateName, description, sortByName, sortByDate).stream().map(dtoMapper::map).collect(Collectors.toList());
+    public List<GiftCertificateDto> getGiftCertificates(String[] tagNames, String giftCertificateName, String description, String sortByName, String sortByDate, Integer page, Integer pageSize) {
+        return giftCertificateDao.getGiftCertificates(tagNames, giftCertificateName, description, sortByName, sortByDate, page, paginator.paginate(page, pageSize, giftCertificateDao.countGiftCertificates())).stream().map(dtoMapper::map).collect(Collectors.toList());
```

Comment:
Please move out some pieces of code from this line. It's very long, not readable.

Reply:
Introduced a new GiftCertificateQueryParameters class to store all the parameters to get in query